### ISSUE TITLE
GRPO silently accepts a reward function that returns the wrong number of rewards

### ISSUE THREAD

**Issue by gowtham-sai-yadav** [huggingface/trl#6533]
### Description

`GRPOTrainer` does not check that a custom reward function returns one reward per completion. In `_generate_and_score_completions`, the reward output is assigned straight into the rewards tensor:

```python
rewards_per_func[:, i] = torch.tensor(output_reward_func, dtype=torch.float32, device=device)
```

Because `rewards_per_func[:, i]` has shape `(num_completions,)`, a reward function that returns a single value produces a `(1,)` tensor that PyTorch **silently broadcasts** to every completion. So a broken reward function corrupts training with no warning. A wrong length that is neither 1 nor `num_completions` instead fails later with an opaque tensor shape error rather than a clear message.

### Reproduction

```python
from datasets import load_dataset
from trl import GRPOConfig, GRPOTrainer

dataset = load_dataset("trl-internal-testing/zen", "standard_prompt_only", split="train")

def bad_reward(completions, **kwargs):
    return [1.0]  # returns 1 reward regardless of the number of completions

args = GRPOConfig(
    output_dir="out", per_device_train_batch_size=3, num_generations=3,
    max_completion_length=8, report_to="none",
)
trainer = GRPOTrainer(
    model="trl-internal-testing/tiny-Qwen2ForCausalLM-2.5",
    reward_funcs=bad_reward, args=args, train_dataset=dataset,
)
trainer.train()
```

**Observed:** training runs with no error and logs the single value broadcast as the reward for every completion (`reward: 1`, `reward_std: 0`).

**Expected:** a clear error telling the user the reward function returned the wrong number of rewards.

### Suggested fix

Add a length check in both the synchronous and async reward paths that raises a `ValueError` naming the reward function and the expected count, before the tensor assignment. Happy to open a PR, I have it ready.
